internal/service/ip/ip_addresses_data_source.go

```diff
@@ -0,0 +1,149 @@
+	Address    types.String `tfsdk:"address"`
+	Family     types.String `tfsdk:"family"`
+	Floating   types.Bool   `tfsdk:"floating"`
+	MAC        types.String `tfsdk:"mac"`
+	PartOfPlan types.Bool   `tfsdk:"part_of_plan"`
+	PTRRecord  types.String `tfsdk:"ptr_record"`
+	Server     types.String `tfsdk:"server"`
+	Zone       types.String `tfsdk:"zone"`
+}
+
+func (d *ipAddressesDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Returns a set of IP Addresses that are associated with the UpCloud account.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "ID of the resource.",
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"addresses": schema.SetNestedBlock{
+				NestedObject: schema.NestedBlockObject{
+					Attributes: map[string]schema.Attribute{
+						"access": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "Is address for utility or public network",
+						},
+						"address": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "An UpCloud assigned IP Address",
+						},
+						"family": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "IP address family",
+						},
+						"floating": schema.BoolAttribute{
+							Computed:            true,
+							MarkdownDescription: "Does the IP Address represents a floating IP Address",
+						},
+						"mac": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "MAC address of server interface to assign address to",
+						},
+						"part_of_plan": schema.BoolAttribute{
+							Computed:            true,
+							MarkdownDescription: "Is the address a part of a plan",
+						},
+						"ptr_record": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "A reverse DNS record entry",
+						},
+						"server": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "The unique identifier for a server",
+						},
+						"zone": schema.StringAttribute{
+							Computed:            true,
+							MarkdownDescription: "Zone of address, required when assigning a detached floating IP address, e.g. `de-fra1`. You can list available zones with `upctl zone list`.",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *ipAddressesDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var diags diag.Diagnostics
+	var data ipAddressesDataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	ipAddresses, err := d.client.GetIPAddresses(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to read ip addresses",
+			utils.ErrorDiagnosticDetail(err),
+		)
+	}
+
+	data.ID = types.StringValue(time.Now().UTC().String())
```

Comment:
I guess the ID could also be something static for data sources that do not have any user defined fields, but maybe better to go through all of those at once 🤔

Reply:
I would keep it as it is for now, but indeed go through them in a major release.